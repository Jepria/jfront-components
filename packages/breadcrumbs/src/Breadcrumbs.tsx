import React, { useState, useRef } from "react"
import { forwardRef } from "@jfront/ui-utils"
import { Popup } from "@jfront/ui-popup"
import styled from "styled-components"
import { Link } from "@jfront/ui-link"

export interface BreadcrumbsProps {
  maxItems?: number
  collapseMethod?: string //"dropdown", "hidden"
  itemsAfterCollapse?: number
  itemsBeforeCollapse?: number
  expandText?: string
  children: React.ReactNode
  separator?: () => React.ReactNode | string | React.ReactNode
}

export const BreadcrumsLink = styled(Link)`
  color: ${(props) => props.theme.breadcrumbs.color};
  text-decoration: none;
  &:hover {
    ${(props) =>
      !!props.href ? "text-decoration: underline;" : "text-decoration: none;"}
  }
`

BreadcrumsLink.defaultProps = {
  theme: {
    fontFamily: "tahoma, arial, helvetica, sans-serif",
    fontSize: {
      medium: "12px",
    },
    link: {
      color: "rgb(21, 66, 139)",
      textDecoration: "underline",
    },
    breadcrumbs: {
      color: "currentColor",
    },
  },
}

const Separator = styled.span`
  box-sizing: border-box;
  font-weight: bold;
  color: ${(props) => props.theme.breadcrumbs.color};
  margin: 0 5px;
`

Separator.defaultProps = {
  theme: {
    breadcrumbs: {
      color: "currentColor",
    },
  },
}

const ExpanderContent = styled.span`
  cursor: pointer;
  &:hover {
    opacity: 0.5;
  }
`

const Dot = styled.span`
  height: 4px;
  width: 4px;
  background: ${(props) => props.theme.breadcrumbs.color};
  border-radius: 50%;
  margin: 0 1px;
  display: inline-block;
`

Dot.defaultProps = {
  theme: {
    breadcrumbs: {
      color: "currentColor",
    },
  },
}

const Nav = styled.nav`
  font-size: ${(props) => props.theme.fontSize.medium};
  font-family: ${(props) => props.theme.fontFamily};
  box-sizing: border-box;
  padding: 0;
  margin: 0;
`

Nav.defaultProps = {
  theme: {
    fontFamily: "tahoma, arial, helvetica, sans-serif",
    fontSize: {
      medium: "12px",
    },
  },
}

const Expander = React.forwardRef<
  HTMLSpanElement,
  React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLSpanElement>,
    HTMLSpanElement
  >
>((props, ref) => {
  return (
    <>
      <ExpanderContent role="button" onClick={props.onClick} ref={ref}>
        <Dot />
        <Dot />
        <Dot />
      </ExpanderContent>
      {props.children}
    </>
  )
})

export const Breadcrumbs = forwardRef<BreadcrumbsProps, "nav">(
  (
    {
      maxItems = 3,
      collapseMethod = "hidden",
      itemsAfterCollapse = 1,
      itemsBeforeCollapse = 1,
      expandText = "Развернуть",
      separator = "/",
      children,
      ...rest
    },
    ref,
  ) => {
    const expanderRef = useRef<HTMLSpanElement>(null)
    const [expanded, setExpanded] = useState(false)

    const addSeparators = (children: any[]) => {
      for (let i = 1; i < children.length; i++) {
        if ((i + 1) % 2 == 0) {
          let element
          if (typeof separator === "string") {
            element = React.createElement(Separator, { key: i }, separator)
          } else if (typeof separator === "function") {
            element = separator()
          } else {
            element = separator
          }
          children.splice(i, 0, element)
        }
      }
      return children
    }

    const render = () => {
      const childrenArray = React.Children.toArray(children)
      if (
        (expanded && collapseMethod === "hidden") ||
        childrenArray.length <= maxItems
      ) {
        return addSeparators(childrenArray)
      } else if (
        collapseMethod === "dropdown" &&
        childrenArray.length > maxItems
      ) {
        return addSeparators([
          childrenArray.splice(0, itemsBeforeCollapse),
          React.createElement(
            Expander,
            {
              title: expandText,
              onClick: () => setExpanded(true),
              key: "expander",
              ref: expanderRef,
            },
            React.createElement(
              Popup,
              {
                style: {
                  padding: "5px",
                  borderRadius: "5px",
                },
                targetElementRef: expanderRef,
                targetRelativePosition: {
                  horizontal: "left",
                  vertical: "bottom",
                },
                visible: expanded,
                onClose: () => setExpanded(false),
              },
              renderHidden(),
            ),
          ),
          childrenArray.splice(childrenArray.length - itemsAfterCollapse),
        ])
      } else {
        return addSeparators([
          childrenArray.splice(0, itemsBeforeCollapse),
          React.createElement(Expander, {
            title: expandText,
            onClick: () => setExpanded(true),
            key: "expander",
          }),
          childrenArray.splice(childrenArray.length - itemsAfterCollapse),
        ])
      }
    }

    const renderHidden = () => {
      const childrenArray = React.Children.toArray(children)
      return childrenArray
        .splice(
          itemsBeforeCollapse,
          childrenArray.length - itemsAfterCollapse - 1,
        )
        .map((child) => {
          if (!React.isValidElement(child)) {
            throw new Error("Invalid React Element")
          }
          const props = (child as any).props
          return React.cloneElement(child as any, {
            ...props,
            style: {
              ...props?.style,
              display: "block",
            },
          })
        })
    }

    return (
      <>
        <Nav {...rest} ref={ref}>
          {render()}
        </Nav>
      </>
    )
  },
)
