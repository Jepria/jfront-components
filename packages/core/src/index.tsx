export * from '@jfront-ui/app'
export * from '@jfront-ui/datepicker'
export * from '@jfront-ui/form'
//export * from '@jfront-ui/grid'
export * from '@jfront-ui/hooks'
//export * from '@jfront-ui/input'
export * from '@jfront-ui/layout'
//export * from '@jfront-ui/pagingbar'
//export * from '@jfront-ui/tabpanel'
//export * from '@jfront-ui/toolbar'
