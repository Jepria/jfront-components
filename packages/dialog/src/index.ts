export * from './SimpleDialog'
export * from './AlertDialog'
export * from './FormDialog'
export * from './ErrorDialog'
export * from './WarningDialog'