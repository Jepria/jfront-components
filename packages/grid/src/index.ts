export * from './common'
export * from './jepgrid'
