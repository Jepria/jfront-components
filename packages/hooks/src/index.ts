export * from "./useOnClickOutside"
export * from "./useRouteMapping"
export * from "./useTree/"
export * from "./useTree/types"
export * from "./useFilter"
