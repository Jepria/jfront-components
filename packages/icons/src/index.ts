export * from "./Add"
export * from "./Delete"
export * from "./Edit"
export * from "./Exclamation"
export * from "./First"
export * from "./Last"
export * from "./Loading"
export * from "./Next"
export * from "./OpenIcon"
export * from "./Prev"
export * from "./Refresh"
export * from "./Save"
export * from "./Search"
export * from "./SettingsSmall"
export * from "./Split"
export * from "./View"
