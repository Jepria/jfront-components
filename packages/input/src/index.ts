export interface InputProps {
  error?: string
  name?: string
  isLoading?: boolean
}

export * from "./SelectInput"
export * from "./TextInput"
export * from "./NumberInput"
export * from "./MaskedInput"
export * from "./MoneyInput"
export * from "./DecimalInput"
export * from "./MoneyInput"
