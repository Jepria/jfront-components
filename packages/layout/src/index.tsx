import styled from 'styled-components';

export const FormContainer = styled.div`
  padding: 5px;
`;

export const VerticalLayout = styled.div`
  width: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
`;

export const HorizontalLayout = styled.div`
  width: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
      -ms-flex-direction: row;
          flex-direction: row;
`;

export const Page = styled.div`
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
`;

export const Header = styled.header`
  box-sizing: border-box;
  height: auto;
  width: 100%;
`;

export const Content = styled.section`
  box-sizing: border-box;
  width: 100%;
  overflow: auto;
  -webkit-box-flex: 1;
      -ms-flex: 1;
          flex: 1;
  -ms-flex-positive: 1;
      flex-grow: 1;
`;

export const Footer = styled.footer`
  box-sizing: border-box;
  height: auto;
  width: 100%;
`;
