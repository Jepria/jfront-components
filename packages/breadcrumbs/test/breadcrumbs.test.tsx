import React from "react"
import { render } from "react-dom"
import { Breadcrumbs } from "../src"
import { Link } from "@jfront/ui-link"
import { fireEvent, screen } from "@testing-library/react"

const container = document.createElement("div")
document.body.appendChild(container)

test("Matches snapshot 1", () => {
  render(
    <Breadcrumbs>
      <Link href="/test1">Test Link1</Link>
      <Link href="/test2">Test Link2</Link>
      <Link href="/test3">Test Link3</Link>
    </Breadcrumbs>,
    container,
  )
  expect(document.body).toMatchSnapshot()
})

test("Matches snapshot 2", () => {
  render(
    <Breadcrumbs>
      <Link href="/test1">Test Link1</Link>
      <Link href="/test2">Test Link2</Link>
      <Link href="/test3">Test Link3</Link>
      <Link href="/test4">Test Link4</Link>
      <Link href="/test5">Test Link5</Link>
    </Breadcrumbs>,
    container,
  )
  expect(document.body).toMatchSnapshot()
})

test("expand click", () => {
  render(
    <Breadcrumbs data-testid="breadcrumbs">
      <Link href="/test1">Test Link1</Link>
      <Link href="/test2">Test Link2</Link>
      <Link data-testid="a" href="/test3">
        Test Link3
      </Link>
      <Link href="/test4">Test Link4</Link>
      <Link>Test Link5</Link>
    </Breadcrumbs>,
    container,
  )
  fireEvent.click(screen.getByRole("button"))
  fireEvent.click(screen.getByTestId("a"))
  expect(window.location.pathname).toEqual("/test3")
})

test("popup click", () => {
  render(
    <Breadcrumbs data-testid="breadcrumbs" collapseMethod="dropdown">
      <Link href="/test1">Test Link1</Link>
      <Link href="/test2">Test Link2</Link>
      <Link data-testid="a" href="/test3">
        Test Link3
      </Link>
      <Link href="/test4">Test Link4</Link>
      <Link>Test Link5</Link>
    </Breadcrumbs>,
    container,
  )
  fireEvent.click(screen.getByRole("button"))
  fireEvent.click(screen.getByTestId("a"))
  expect(window.location.pathname).toEqual("/test3")
})
