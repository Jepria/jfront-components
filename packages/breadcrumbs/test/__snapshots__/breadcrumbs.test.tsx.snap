// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1 1`] = `
<body>
  <div>
    <nav
      class="sc-AxhUy fNHVAv"
    >
      <a
        class="sc-AxheI cshSQi"
        href=""
      >
        Test Link1
      </a>
      <span
        class="sc-AxirZ fdRIax"
      >
        /
      </span>
      <a
        class="sc-AxheI cshSQi"
        href=""
      >
        Test Link2
      </a>
      <span
        class="sc-AxirZ fdRIax"
      >
        /
      </span>
      <a
        class="sc-AxheI cshSQi"
        href=""
      >
        Test Link3
      </a>
    </nav>
  </div>
</body>
`;

exports[`Matches snapshot 2 1`] = `
<body>
  <div>
    <nav
      class="sc-AxhUy fNHVAv"
    >
      <a
        class="sc-AxheI cshSQi"
        href=""
      >
        Test Link1
      </a>
      <span
        class="sc-AxirZ fdRIax"
      >
        /
      </span>
      <span
        class="sc-AxiKw ehZmPD"
        role="button"
      >
        <span
          class="sc-AxhCb hgbnDK"
        />
        <span
          class="sc-AxhCb hgbnDK"
        />
        <span
          class="sc-AxhCb hgbnDK"
        />
      </span>
      <span
        class="sc-AxirZ fdRIax"
      >
        /
      </span>
      <a
        class="sc-AxheI cshSQi"
        href=""
      >
        Test Link5
      </a>
    </nav>
  </div>
</body>
`;
